File: crates/lenke-core/src/algo/centrality.rs
//! Betweenness and closeness centrality — two shortest-path-based measures over the
//! directed graph (out-edges, optionally one `edgeLabel`, optionally weighted by a
//! `weightProperty`). Both are O(V·E) unweighted / O(V·(E + V log V)) weighted: one
//! single-source shortest-path pass per vertex, so they do NOT scale to very large
//! graphs — see the README perf note.
//!
//! - **Betweenness** (Brandes' algorithm): for each vertex `v`, the sum over all
//!   ordered pairs `(s,t)` of the fraction of shortest `s→t` paths that pass through
//!   `v`. Directed and UNNORMALIZED (no `1/((n-1)(n-2))` scaling). Endpoints score 0
//!   for their own pairs.
//! - **Closeness**: `1 / Σ_t d(s,t)` over every `t` reachable from `s`
//!   (t ≠ s). UNNORMALIZED (not `(reachable-1) / Σ d`). A vertex that reaches nothing
//!   scores 0.
//!
//! Cross-engine bit-identity: both engines build each vertex's out-adjacency by
//! scanning **all edges once in global edge-insertion order** (a CSR, exactly like
//! `pagerank`), so the BFS queue / Dijkstra settle order — and therefore Brandes'
//! stack + predecessor lists — are identical, making every f64 dependency
//! accumulation `δ[v] += (σ[v]/σ[w])·(1+δ[w])` add its terms in the same sequence as
//! the TS mirror. `σ` (shortest-path counts) are integers summed exactly; the
//! closeness distance sum is taken in vertex-insertion order. The Dijkstra frontier
//! breaks ties by `(dist, vertex index)`, matching the TS heap.

use std::collections::{BinaryHeap, VecDeque};

use super::AlgoConfig;
use crate::graph::{Graph, Value};

/// A compressed-sparse-row out-adjacency: `nbr[off[v]..off[v+1]]` are v's
/// out-neighbours (with parallel edge weights in `w`), built in global
/// edge-insertion order so both engines traverse a vertex's edges identically.
struct OutCsr {
    off: Vec<usize>,
    nbr: Vec<u32>,
    w: Vec<f64>,
    weighted: bool,
}

/// Build the out-adjacency CSR, scanning all live edges once in insertion order.
fn build_csr(graph: &Graph, cfg: &AlgoConfig) -> OutCsr {
    let slots = graph.n;
    // Some(None) = every type; Some(Some(t)) = one type; None = named-but-unknown
    // type → no edges match (the graph is treated as edgeless).
    let etype = cfg.etype(graph);
    // ANY of the edge's labels — `e_type` alone is only its FIRST, which made
    // every algorithm here see an edgeless graph when the filtered label was
    // stored second.
    let type_ok = |ei: usize| match etype {
        Some(Some(t)) => graph.edge_has_label(ei as u32, t),
        Some(None) => true,
        None => false,
    };
    let weights: Option<Vec<f64>> = cfg
        .weight_property
        .as_deref()
        .map(|k| super::edge_weights(graph, k));
    let weight_of = |ei: usize| -> f64 { weights.as_ref().map_or(1.0, |w| w[ei]) };

    let mut off = vec![0usize; slots + 1];
    for ei in 0..graph.edge_slots() {
        if graph.is_edge_live(ei as u32) && type_ok(ei) {
            off[graph.e_src[ei] as usize + 1] += 1;
        }
    }
    for v in 0..slots {
        off[v + 1] += off[v];
    }
    let mut nbr = vec![0u32; off[slots]];
    let mut w = vec![0f64; off[slots]];
    let mut cursor = off[..slots].to_vec();
    for ei in 0..graph.edge_slots() {
        if !graph.is_edge_live(ei as u32) || !type_ok(ei) {
            continue;
        }
        let src = graph.e_src[ei] as usize;
        let pos = cursor[src];
        cursor[src] += 1;
        nbr[pos] = graph.e_dst[ei];
        w[pos] = weight_of(ei);
    }
    OutCsr {
        off,
        nbr,
        w,
        weighted: weights.is_some(),
    }
}

/// A Dijkstra frontier entry ordered as a min-heap on `(dist, idx)` (see
/// `shortest_path::State` — same tie-break so both engines settle identically).
struct State {
    dist: f64,
    idx: u32,
}
impl PartialEq for State {
    fn eq(&self, other: &Self) -> bool {
        self.dist == other.dist && self.idx == other.idx
    }
}
impl Eq for State {}
impl Ord for State {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        other
            .dist
            .partial_cmp(&self.dist)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then_with(|| other.idx.cmp(&self.idx))
    }
}
impl PartialOrd for State {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

/// The per-source shortest-path DAG Brandes needs: the settle/visit `stack` (in
/// non-decreasing distance order), shortest-path counts `sigma`, and predecessor
/// lists `pred` (each in global edge-insertion order).
struct Sssp {
    stack: Vec<u32>,
    sigma: Vec<f64>,
    pred: Vec<Vec<u32>>,
    dist: Vec<f64>,
}

/// Single-source shortest paths from `s` over the CSR — unweighted BFS layers or
/// weighted Dijkstra — recording the Brandes bookkeeping. Explores neighbours in
/// CSR (edge-insertion) order, so the stack and predecessor lists are deterministic.
fn sssp(csr: &OutCsr, slots: usize, s: u32) -> Sssp {
    let mut sigma = vec![0f64; slots];
    let mut pred: Vec<Vec<u32>> = vec![Vec::new(); slots];
    let mut stack: Vec<u32> = Vec::new();
    let mut dist = vec![f64::INFINITY; slots];
    sigma[s as usize] = 1.0;
    dist[s as usize] = 0.0;

    if csr.weighted {
        let mut settled = vec![false; slots];
        let mut heap = BinaryHeap::new();
        heap.push(State { dist: 0.0, idx: s });
        while let Some(State { idx: v, .. }) = heap.pop() {
            if settled[v as usize] {
                continue;
            }
            settled[v as usize] = true;
            stack.push(v);
            let dv = dist[v as usize];
            for j in csr.off[v as usize]..csr.off[v as usize + 1] {
                let to = csr.nbr[j];
                let nd = dv + csr.w[j];
                if nd < dist[to as usize] {
                    dist[to as usize] = nd;
                    sigma[to as usize] = sigma[v as usize];
                    pred[to as usize] = vec![v];
                    heap.push(State { dist: nd, idx: to });
                } else if nd == dist[to as usize] {
                    sigma[to as usize] += sigma[v as usize];
                    pred[to as usize].push(v);
                }
            }
        }
    } else {
        let mut queue = VecDeque::new();
        queue.push_back(s);
        while let Some(v) = queue.pop_front() {
            stack.push(v);
            let dv = dist[v as usize];
            for j in csr.off[v as usize]..csr.off[v as usize + 1] {
                let to = csr.nbr[j];
                if dist[to as usize].is_infinite() {
                    dist[to as usize] = dv + 1.0;
                    queue.push_back(to);
                }
                if dist[to as usize] == dv + 1.0 {
                    sigma[to as usize] += sigma[v as usize];
                    pred[to as usize].push(v);
                }
            }
        }
    }

    Sssp {
        stack,
        sigma,
        pred,
        dist,
    }
}

/// Betweenness centrality (Brandes). Accumulates each source's dependencies in
/// reverse-stack (non-increasing distance) order — a fixed, deterministic order —
/// so the per-vertex f64 sum is byte-identical to the TS mirror.
pub fn betweenness(graph: &Graph, cfg: &AlgoConfig) -> Vec<(u32, Value)> {
    let slots = graph.n;
    let csr = build_csr(graph, cfg);
    let verts: Vec<u32> = graph.vertex_indices().collect();
    let n = verts.len();

    // Source set: exact = every vertex; approximate = a deterministic evenly-spaced
    // sample of `pivots` vertices (indices `i*n/k`), so both engines pick the SAME
    // sources → byte-identical estimate. A sampled run scales the summed
    // dependencies by n/k (the Brandes–Pich estimator).
    let sources: Vec<u32> = match cfg.pivots {
        Some(k) if (k as usize) < n && k > 0 => {
            let k = k as usize;
            (0..k).map(|i| verts[i * n / k]).collect()
        }
        _ => verts.clone(),
    };

    let mut cb = vec![0f64; slots];
    for &s in &sources {
        let sp = sssp(&csr, slots, s);
        let mut delta = vec![0f64; slots];
        // Pop in reverse visit order (non-increasing distance): each w's dependency
        // is final before it flows back to its predecessors.
        for &w in sp.stack.iter().rev() {
            let coeff = 1.0 + delta[w as usize];
            for &v in &sp.pred[w as usize] {
                delta[v as usize] += (sp.sigma[v as usize] / sp.sigma[w as usize]) * coeff;
            }
            if w != s {
                cb[w as usize] += delta[w as usize];
            }
        }
    }

    // Scale a sampled run up to a full-graph estimate. (Exact runs use every
    // source, so `sources.len() == n` and the scale is 1.)
    if sources.len() < n {
        let scale = n as f64 / sources.len() as f64;
        for v in &mut cb {
            *v *= scale;
        }
    }

    graph
        .vertex_indices()
        .map(|v| (v, Value::Num(cb[v as usize])))
        .collect()
}

/// Closeness centrality: `1 / Σ_t d(s,t)` over reachable `t ≠ s`, summed in
/// vertex-insertion order (0 when nothing is reachable). Unnormalized.
pub fn closeness(graph: &Graph, cfg: &AlgoConfig) -> Vec<(u32, Value)> {
    let slots = graph.n;
    let csr = build_csr(graph, cfg);

    graph
        .vertex_indices()
        .map(|s| {
            let sp = sssp(&csr, slots, s);
            // Sum finite distances in vertex-insertion order (the source's own 0
            // contributes nothing). INF (unreachable) is excluded.
            let mut sum = 0.0f64;
            for v in graph.vertex_indices() {
                let d = sp.dist[v as usize];
                if d.is_finite() {
                    sum += d;
                }
            }
            let c = if sum == 0.0 { 0.0 } else { 1.0 / sum };
            (s, Value::Num(c))
        })
        .collect()
}

File: packages/core/src/algorithms/centrality.ts
import { ErrorCode, LenkeError } from '@lenke/errors';

import type { Edge } from '../core/Edge.js';
import type { Graph } from '../core/Graph.js';
import { type AlgorithmGen, defineAlgorithm, materializeVertices, YIELD_EVERY } from './async.js';
import type { AlgorithmConfig, AlgorithmRow } from './types.js';

/** A betweenness / closeness result row: `{ node, centrality }`. */
export type CentralityRow = AlgorithmRow<'centrality', number>;

/**
 * A compressed-sparse-row out-adjacency built by scanning ALL edges once in global
 * edge-insertion order (matching native `build_csr`), so both engines traverse a
 * vertex's out-edges in the exact same sequence — the key to byte-identical BFS /
 * Dijkstra exploration and therefore identical Brandes dependency accumulation.
 */
type OutCsr = { off: Int32Array; nbr: Int32Array; w: Float64Array; weighted: boolean };

/** Build the out-adjacency CSR, yielding at checkpoints so the O(E) setup is frame-safe. */
const buildCsr = function* (
  graph: Graph,
  order: readonly { id: string }[],
  index: Map<string, number>,
  config: AlgorithmConfig,
): Generator<void, OutCsr, void> {
  const { edgeLabel, weightProperty } = config;
  const n = order.length;
  const weighted = weightProperty !== undefined;
  const typeOk = (edge: Edge): boolean => edgeLabel === undefined || edge.labels.has(edgeLabel);
  const weightOf = (edge: Edge): number => {
    if (weightProperty === undefined) {
      return 1;
    }

    const x = edge.getProperty(weightProperty);

    return typeof x === 'number' ? x : 0;
  };

  // Weighted closeness/betweenness run Dijkstra, which requires NON-NEGATIVE weights —
  // a negative edge never settles and a NaN strands the search, leaving the distances
  // undefined. Reject any negative/NaN weight up front (matching shortestPath + the
  // native engine), whether or not it lies on a matched-label reachable path.
  if (weighted) {
    for (const edge of graph.edges) {
      const x = edge.getProperty(weightProperty);

      if (typeof x === 'number' && (Number.isNaN(x) || x < 0)) {
        throw new LenkeError(
          `\`weightProperty\` (${weightProperty}) must hold non-negative numbers — ` +
            `Dijkstra does not admit negative weights`,
          { code: ErrorCode.InvalidValue },
        );
      }
    }
  }

  let sinceYield = 0;
  const off = new Int32Array(n + 1);

  for (const edge of graph.edges) {
    if (typeOk(edge)) {
      off[index.get(edge.from.id)! + 1]++;
    }

    if (++sinceYield >= YIELD_EVERY) {
      sinceYield = 0;

      yield;
    }
  }

  for (let v = 0; v < n; v++) {
    off[v + 1] += off[v];
  }

  const nbr = new Int32Array(off[n]);
  const w = new Float64Array(off[n]);
  const cursor = off.slice(0, n);

  for (const edge of graph.edges) {
    if (!typeOk(edge)) {
      continue;
    }

    const src = index.get(edge.from.id)!;
    const pos = cursor[src]++;
    nbr[pos] = index.get(edge.to.id)!;
    w[pos] = weightOf(edge);

    if (++sinceYield >= YIELD_EVERY) {
      sinceYield = 0;

      yield;
    }
  }

  return { off, nbr, w, weighted };
};

/** A binary min-heap entry `[distance, vertexIndex]`, ordered by distance then index. */
type HeapItem = [number, number];

const less = (a: HeapItem, b: HeapItem): boolean => a[0] < b[0] || (a[0] === b[0] && a[1] < b[1]);

const heapPush = (heap: HeapItem[], item: HeapItem): void => {
  heap.push(item);
  let i = heap.length - 1;

  while (i > 0) {
    const parent = (i - 1) >> 1;

    if (!less(heap[i], heap[parent])) {
      break;
    }

    [heap[i], heap[parent]] = [heap[parent], heap[i]];
    i = parent;
  }
};

const heapPop = (heap: HeapItem[]): HeapItem => {
  const [top] = heap;
  const last = heap.pop()!;

  if (heap.length > 0) {
    heap[0] = last;
    let i = 0;

    for (;;) {
      const l = 2 * i + 1;
      const r = l + 1;
      let smallest = i;

      if (l < heap.length && less(heap[l], heap[smallest])) {
        smallest = l;
      }

      if (r < heap.length && less(heap[r], heap[smallest])) {
        smallest = r;
      }

      if (smallest === i) {
        break;
      }

      [heap[i], heap[smallest]] = [heap[smallest], heap[i]];
      i = smallest;
    }
  }

  return top;
};

/** The per-source shortest-path DAG (Brandes bookkeeping): visit stack, path counts,
 *  predecessor lists, and distances — mirroring native `Sssp`. */
type Sssp = { stack: number[]; sigma: Float64Array; pred: number[][]; dist: Float64Array };

/**
 * Single-source shortest paths from `s` over the CSR — unweighted BFS layers or
 * weighted Dijkstra — recording the stack (settle order), `sigma` (path counts) and
 * predecessor lists (in CSR / edge-insertion order). The Dijkstra frontier ties on
 * `(dist, vertex index)`, matching native, so the stack + predecessors are identical.
 */
const sssp = (csr: OutCsr, n: number, s: number): Sssp => {
  const sigma = new Float64Array(n);
  const pred: number[][] = Array.from({ length: n }, () => []);
  const stack: number[] = [];
  const dist = new Float64Array(n).fill(Infinity);
  sigma[s] = 1;
  dist[s] = 0;

  if (csr.weighted) {
    const settled = new Uint8Array(n);
    const heap: HeapItem[] = [[0, s]];

    while (heap.length > 0) {
      const [, v] = heapPop(heap);

      if (settled[v]) {
        continue;
      }

      settled[v] = 1;
      stack.push(v);
      const dv = dist[v];

      for (let j = csr.off[v]; j < csr.off[v + 1]; j++) {
        const to = csr.nbr[j];
        const nd = dv + csr.w[j];

        if (nd < dist[to]) {
          dist[to] = nd;
          sigma[to] = sigma[v];
          pred[to] = [v];
          heapPush(heap, [nd, to]);
        } else if (nd === dist[to]) {
          sigma[to] += sigma[v];
          pred[to].push(v);
        }
      }
    }
  } else {
    const queue: number[] = [s];
    let head = 0;

    while (head < queue.length) {
      const v = queue[head++];
      stack.push(v);
      const dv = dist[v];

      for (let j = csr.off[v]; j < csr.off[v + 1]; j++) {
        const to = csr.nbr[j];

        if (dist[to] === Infinity) {
          dist[to] = dv + 1;
          queue.push(to);
        }

        if (dist[to] === dv + 1) {
          sigma[to] += sigma[v];
          pred[to].push(v);
        }
      }
    }
  }

  return { stack, sigma, pred, dist };
};

export const betweennessGen = function* (
  config: AlgorithmConfig,
  graph: Graph,
): AlgorithmGen<CentralityRow> {
  const order = yield* materializeVertices(graph);
  const n = order.length;

  if (n === 0) {
    return [];
  }

  const index = new Map<string, number>();
  order.forEach((v, i) => index.set(v.id, i));

  const csr = yield* buildCsr(graph, order, index, config);
  const cb = new Float64Array(n);
  let sinceYield = 0;

  // Source set: exact = every vertex; approximate (config.pivots) = a deterministic
  // evenly-spaced sample of `pivots` sources (indices `i*n/k`), identical to native,
  // scaled below by n/k (the Brandes–Pich estimator).
  const { pivots } = config;
  const sources: number[] =
    pivots !== undefined && pivots < n && pivots > 0
      ? Array.from({ length: pivots }, (_, i) => Math.floor((i * n) / pivots))
      : Array.from({ length: n }, (_, i) => i);

  for (const s of sources) {
    const sp = sssp(csr, n, s);
    const delta = new Float64Array(n);

    // Pop in reverse visit order (non-increasing distance): each w's dependency is
    // final before it flows back to its predecessors. Same order as native, so the
    // per-vertex f64 accumulation is byte-identical.
    for (let k = sp.stack.length - 1; k >= 0; k--) {
      const w = sp.stack[k];
      const coeff = 1 + delta[w];

      for (const v of sp.pred[w]) {
        delta[v] += (sp.sigma[v] / sp.sigma[w]) * coeff;
      }

      if (w !== s) {
        cb[w] += delta[w];
      }
    }

    if (++sinceYield >= YIELD_EVERY) {
      sinceYield = 0;

      yield;
    }
  }

  // Scale a sampled run up to a full-graph estimate (exact runs use every source).
  if (sources.length < n) {
    const scale = n / sources.length;

    for (let i = 0; i < n; i++) {
      cb[i] *= scale;
    }
  }

  const rows: CentralityRow[] = new Array(n);

  for (let i = 0; i < n; i++) {
    const v = order[i];
    const score = cb[i];

    if (config.writeProperty !== undefined) {
      v.setProperty(config.writeProperty, score);
    }

    rows[i] = { node: v.id, centrality: score };
  }

  return rows;
};

export const closenessGen = function* (
  config: AlgorithmConfig,
  graph: Graph,
): AlgorithmGen<CentralityRow> {
  const order = yield* materializeVertices(graph);
  const n = order.length;

  if (n === 0) {
    return [];
  }

  const index = new Map<string, number>();
  order.forEach((v, i) => index.set(v.id, i));

  const csr = yield* buildCsr(graph, order, index, config);
  const rows: CentralityRow[] = new Array(n);
  let sinceYield = 0;

  for (let s = 0; s < n; s++) {
    const sp = sssp(csr, n, s);

    // Sum finite distances in vertex-insertion order (source's own 0 adds nothing;
    // unreachable Infinity excluded) — same order as native, so identical.
    let sum = 0;

    for (let v = 0; v < n; v++) {
      const d = sp.dist[v];

      if (Number.isFinite(d)) {
        sum += d;
      }
    }

    const centrality = sum === 0 ? 0 : 1 / sum;
    const vtx = order[s];

    if (config.writeProperty !== undefined) {
      vtx.setProperty(config.writeProperty, centrality);
    }

    rows[s] = { node: vtx.id, centrality };

    if (++sinceYield >= YIELD_EVERY) {
      sinceYield = 0;

      yield;
    }
  }

  return rows;
};

/**
 * Betweenness centrality (Brandes' algorithm) over the directed graph (out-edges,
 * optionally one `edgeLabel`, optionally weighted by `weightProperty`). Each
 * vertex's score is the sum over all ordered pairs `(s,t)` of the fraction of
 * shortest `s→t` paths through it — directed and UNNORMALIZED. Dependencies
 * accumulate in a fixed reverse-visit order, so scores are byte-identical to the
 * native engine. **O(V·E)** (unweighted) — does not scale to very large graphs.
 * Data-last dual-form: `betweenness(config, graph)` / `betweenness(config)(graph)`.
 */
export const betweenness = defineAlgorithm(betweennessGen);

/**
 * Closeness centrality — `1 / Σ_t d(s,t)` over every reachable `t ≠ s`
 * (UNNORMALIZED; a vertex reaching nothing scores 0). Follows out-edges, optionally
 * one `edgeLabel`, unweighted BFS or (with `weightProperty`) Dijkstra. The distance
 * sum is taken in vertex-insertion order, so scores are byte-identical to native.
 * **O(V·E)** — does not scale to very large graphs. Data-last dual-form:
 * `closeness(config, graph)` / `closeness(config)(graph)`.
 */
export const closeness = defineAlgorithm(closenessGen);
